File: exercise/743/solution.cpp
class Graph {
    vector<unordered_map<int, int>> edges;
public:
    Graph(int n): edges(n + 1) {}

    void insert(int from, int to, int time) {
        edges[from][to] = time;
    }

    int dijkstra(int s) {
        vector<int> dis(edges.size(), 1000000);
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> queue;
        unordered_set<int> reached;
        queue.push(make_pair(0, s));
        dis[s] = 0;
        while (!queue.empty()) {
            pair<int, int> now = queue.top();
            queue.pop();
            if (reached.count(now.second) != 0)
                continue;
            reached.insert(now.second);
            for (const pair<int, int> &it: edges[now.second]) {
                if (dis[it.first] > dis[now.second] + it.second) {
                    dis[it.first] = dis[now.second] + it.second;
                    queue.push(make_pair(dis[it.first], it.first));
                }
            }
        }
        int result = 0;
        for (int i = 1; i < edges.size(); i++) {
            if (dis[i] == 1000000) {
                return -1;
            } else if (dis[i] > result) {
                result = dis[i];
            }
        }
        return result;
    }
};

class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int N, int K) {
        Graph g(N);
        for (int i = 0; i < times.size(); i++) {
            g.insert(times[i][0], times[i][1], times[i][2]);
        }
        return g.dijkstra(K);
    }
};

File: exercise/743/solution.py
from heapq import *

class Graph:

    def __init__(self, size):
        self._graph = [{} for _ in range(0, size + 1)]

    def add(self, f, t, w):
        self._graph[f][t] = w

    def dijkstra(self, v):
        candidates = []
        reached = set()
        all_dis = [100000000 for _ in range(len(self._graph))]
        all_dis[v] = 0
        heappush(candidates, (0, v))
        while len(candidates) > 0:
            dis, t = heappop(candidates)
            if t in reached:
                continue
            reached.add(t)
            for node, l in self._graph[t].items():
                if dis + l < all_dis[node]:
                    all_dis[node] = dis + l
                    heappush(candidates, (all_dis[node], node))
        result = 0
        for dis in all_dis[1:]:
            if dis > result:
                result = dis
        if result == 100000000:
            return -1
        else:
            return result



class Solution:
    def networkDelayTime(self, times: List[List[int]], N: int, K: int) -> int:
        g = Graph(N)
        for edge in times:
            g.add(edge[0], edge[1], edge[2])
        return g.dijkstra(K)
